integration-test/Analysis/FixtureUtils.hs

```diff
@@ -0,0 +1,178 @@
+  runExecIO,
+ )
+import Effect.Logger (LoggerC, Severity (SevDebug), withDefaultLogger)
+import Effect.ReadFS (ReadFSIOC, runReadFSIO)
+import Network.HTTP.Req (
+  GET (GET),
+  NoReqBody (NoReqBody),
+  defaultHttpConfig,
+  reqBr,
+  runReq,
+  useHttpsURI,
+  Url,
+ )
+import Network.HTTP.Req.Conduit (responseBodySource)
+import Path
+import Path.IO qualified as PIO
+import Text.URI (mkURI)
+import Types
+import System.Directory.Internal.Prelude (hClose, Handle)
+
+analysisIntegrationCaseFixtureDir :: Path Rel Dir
+analysisIntegrationCaseFixtureDir = [reldir|integration-test/artifacts/|]
+
+-- | Represents analysis test fixture.
+data AnalysisTestFixture a = AnalysisTestFixture
+  { testName :: Text
+  , discover :: Path Abs Dir -> TestC IO [DiscoveredProject a]
+  , environment :: FixtureEnvironment
+  , buildCmd :: Maybe (Command, Path Rel Dir)
+  , artifact :: FixtureArtifact
+  }
+
+
+-- | Fixture Environment to mimic when executing commands.
+data FixtureEnvironment
+  = NixEnvRawExpression Text
+  | NixEnvSimpleConfig [Text]
+  | LocalEnvironment
+  deriving (Show, Eq, Ord)
+
+-- | Artifact to download and use for the test.
+data FixtureArtifact = FixtureArtifact
+  { tarGzFileUrl :: Text
+  , extractAt :: Path Rel Dir
+  , scopedDir :: Maybe (Path Rel Dir)
+  }
+  deriving (Show, Eq, Ord)
+
+type TestC m a = ExecIOC (ReadFSIOC (DiagnosticsC (LoggerC ((ReaderC AnalyzeExperimentalPreferences) (FinallyC m))))) a
+
+testRunnerWithLogger :: (Has (Lift IO) sig m) => TestC m a -> FixtureEnvironment -> m (Either FailureBundle a)
+testRunnerWithLogger f env =
+  f
+    & runExecIOWithinEnv env
+    & runReadFSIO
+    & runDiagnostics
+    & withDefaultLogger SevDebug
+    & runReader (AnalyzeExperimentalPreferences Nothing)
+    & runFinally
+
+runExecIOWithinEnv :: (Has (Lift IO) sig m) => FixtureEnvironment -> ExecIOC m a -> m a
+runExecIOWithinEnv conf = interpret $ \case
+  Exec dir cmd -> sendIO $ runExecIO $ sendSimple (Exec dir $ decorateCmdWith conf cmd)
+
+decorateCmdWith :: FixtureEnvironment -> Command -> Command
+decorateCmdWith (NixEnvSimpleConfig pkgs) cmd =
+  Command
+    { cmdName = "nix-shell"
+    , cmdArgs = ["-p"] <> pkgs <> ["--run"] <> [cmdName cmd <> " " <> Text.intercalate " " (cmdArgs cmd)]
+    , cmdAllowErr = cmdAllowErr cmd
+    }
+decorateCmdWith (NixEnvRawExpression nixExpression) cmd =
+  Command
+    { cmdName = "nix-shell"
+    , cmdArgs = ["-I"] <> [nixExpression] <> ["--run"] <> [cmdName cmd <> " " <> Text.intercalate " " (cmdArgs cmd)]
+    , cmdAllowErr = cmdAllowErr cmd
+    }
+decorateCmdWith LocalEnvironment cmd = cmd
+
+-- --------------------------------
+-- Analysis fixture test runner
+
+performDiscoveryAndAnalyses :: (Has (Lift IO) sig m, AnalyzeProject a, MonadFail m) => AnalysisTestFixture a -> m [(DiscoveredProject a, DependencyResults)]
+performDiscoveryAndAnalyses AnalysisTestFixture{..} = do
+  sendIO $ runCmd environment buildCmd
+  absoluteArtifactDir <- absoluteDirOf (extractAt artifact)
```

Comment:
Should these artifacts be getting extracted to a temp directory? I don't feel strongly either way, but I do think temp directories are a pattern we follow with downloaded files in this integration test suite and extracted projects in the CLI v1 integration test suite. Is there a benefit to not having them in a temp directory?

Reply:
Mostly since it makes it easier to debug, I will change it to temp directory once all the tests have been added (if that's ok with you)